### PR TITLE
add faster calc max short/long

### PR DESCRIPTION
**Pull request by jrhea**
get_max_long and get_max_short are now analytical functions.

### CODE REVIEW

File: src/elfpy/pricing_models/base.py [hunk 1]
```diff
@@ -382,72 +382,16 @@ def get_max_long(
...
-            return 0, 0
-        last_maybe_max_long = 0, 0
-        bond_percent = 1
-        num_iters = 25
-        for step_size in [1 / (2 ** (x + 1)) for x in range(num_iters)]:
-            # Compute the amount of base needed to purchase the specified amount
-            # of bonds.
-            trade_result = self.calc_in_given_out(
-                out=Quantity(amount=available_bonds * bond_percent, unit=TokenType.PT),
+        epsilon = 1e-7
```
Comment (sentilesdal) [inline src/elfpy/pricing_models/base.py:385]: this to account for rounding errors?

**Comment by wakamex** [comment]
how does this work? writing out what I see:

get_max_long:
```
τ = 1-t
k = z^τ + (2y*z)^τ
base = (k - (z+y)^τ)^(1/τ) - y
bonds = calc_out_given_in(base)
```

get_max_short
```
τ = 1-t
k = z^τ + (2y+z)^τ
bonds = k^(1/τ) - 2*y - z
base = calc_out_given_in(bonds)
```

is the derivation something like this?
```
k = k_new
z^τ + (2y+z)^τ = (z+d_z)^τ + (y+d_y+y+z)^τ
in the short case, d_z = -z (you're extracting all the shares from the market)
= (z-z)^τ + (2y+d_y+z)^τ
take ln() of both sides:
ln(z^τ + (2y+z)^τ) = ln((2y+d_y+z)^τ)
τ*ln(z^τ + (2y+z)) = τ*ln(2y+d_y+z)
ln(z^τ + (2y+z)) = ln(2y+d_y+z)
take e^x of both sides:
e^(ln(z^τ + (2y+z))) = e^(ln(2y+d_y+z))
z^τ + (2y+z) = 2y+d_y+z
z^τ + 2y - 2y + z - z = d_y
d_y = z^τ
```

File: src/elfpy/pricing_models/base.py [hunk 2]
```diff
@@ -382,72 +382,16 @@ def get_max_long(
...
+        k = pow(market_state.share_reserves,1-time_remaining.stretched_time) + pow(2*market_state.bond_reserves+market_state.share_reserves,1-time_remaining.stretched_time)
```
Comment (sentilesdal) [inline src/elfpy/pricing_models/base.py:386]: don't we need initial_share_price and share_price on the left hand side of this?
`k = (share_price * initial_share_price) * pow(initial_share_price * market_state.share_reserves, 1 - time_remaining.stretched_time)...`

Comment (jrhea) [reply src/elfpy/pricing_models/base.py:386]: ya

File: src/elfpy/pricing_models/base.py (hunk 1 above)
Comment (jrhea) [reply src/elfpy/pricing_models/base.py]: removed

**Review by dpaiton — APPROVED** [review]
lgtm

File: src/elfpy/pricing_models/base.py [hunk 3]
```diff
@@ -375,79 +378,24 @@ def get_max_long(
...
-                    bond_percent -= step_size
-                else:
-                    last_maybe_max_long = (maybe_max_long, d_bonds)
-                    if bond_percent == 1:
-                        return last_maybe_max_long
-                    bond_percent += step_size
-
-        return last_maybe_max_long
+            ).breakdown.with_fee
+            - market_state.base_buffer / market_state.share_price
```
Comment (dpaiton) [inline src/elfpy/pricing_models/base.py:391]: Is this -c * z a hyperdrive thing? this change is in `base` -- would that break for yieldspace?

**Pull request merged**